File: 2021/s3.cpp
#include <bits/stdc++.h>

using namespace std;

vector<int> p, w, d;
int n;

long long move(int x){ // Function that takes the position of the concert and returns time takes for everyone to hear it
    long long time= 0;
    for(int a=0;a<n;++a){ // Loop through each friend
        int rnl = p[a]-d[a]; // Left and right hearing range
        int rnr = p[a]+d[a];

        // Check if the friend can already hear concert
        if(x >= rnl and x <= rnr){
            continue;
        }
        int close = rnr;
        if(abs(x-rnl) < abs(x-rnr)){
            close = rnl;
        }

        time  += (long long)w[a] * abs(x-close);
    }
    return time;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    

    // Get input
    cin >> n;
    
    p.resize(n);
    w.resize(n);
    d.resize(n);

    int l = 1e9;
    int r = 0;

    for(int a=0;a<n;++a){
        cin >> p[a] >> w[a] >> d[a];
        l = min(l, p[a]);
        r = max(r, p[a]);
    }


    // Ternary Search for Function Minimum

    long long best = 1e18;
    for(int a=0;a<60;++a){
        int delta = (r-l)/3;
        int m1 = l+delta;
        int m2 = r-delta;
        long long x1 = move(m1);
        long long x2 = move(m2);
        best = min(x1, best);
        best = min(x2, best);
        if(x1 > x2){
            l = m1;
        }
        else{
            r = m2;
        }
    }

    cout << best  << endl;
    
    return 0;
}

File: 2021/s3.py
# Read C++ code for comments (same idea) and watch youtube video for more details


import sys

input = sys.stdin.readline


p = []
w = []
d = []

n = int(input())

def move(c: int) -> int:
    time = 0
    for a in range(n):
        leftRange = p[a]-d[a]
        rightRange = p[a]+d[a]

        if c >= leftRange and c<= rightRange:
            continue

        closest = rightRange
        if abs(c-leftRange) < abs(c-rightRange):
            closest = leftRange
        time += w[a]*abs(closest-c)
    return time


if __name__ == '__main__':
    l = 1e9
    r = 0
    for a in range(n):
        data = input().strip()
        data = data.split()
        p.append(int(data[0]))
        w.append(int(data[1]))
        d.append(int(data[2]))
        l = min(l, p[a])
        r = max(r, p[a])

    best = 1e18
    for _ in range(60):
        delta = int((r-l)/3)
        m1 = l+delta
        m2 = r-delta
        x1 = move(m1)
        x2 = move(m2)
        best = min(best, x1)
        best = min(best, x2)
        if x1 < x2:
            r = m2
        else:
            l = m1
    
    print(best)

    
